Support phe algorithm Twisted-EC-ElGamal.

wa5i opened the pull request:
<!--
发起一个新Pull Request的通用原则：

1. 在PR的Description中明确说明此PR的作用
2. 如果此PR和某个Issue有关联，则需要进行显式关联，例如在PR中写明：Fixes #xxxx
3. 完成下列checklist的检查
-->

##### Checklist
<!-- 基于你的PR的实际情况移除不适用的项目。其他完成的项目修改[ ]为[x]. -->
- [ ] 在 https://yuque.com/tsdoc 增加或更新了必要的文档
- [x] 增加或更新了必要的测试用例
- [x] 对于重要修改，更新了CHANGES文件
- [x] 当前修改存在对已有API参数或返回值的改变
- [ ] 当前修改存在对旧版本功能的兼容性改变（如网络协议或密码算法）

wa5i committed:
Support phe algorithm Twisted-EC-ElGamal.

dongbeiouba commented on crypto/ec/ec_elgamal_crypt.c:
```diff
@@ -0,0 +1,477 @@
+int EC_POINT_from_string(const EC_GROUP *group, EC_POINT *r,
+                         const unsigned char *str, size_t len)
+{
+	int ret = 0;
```
缩进问题

wa5i replied:
> 缩进问题

已修复

dongbeiouba commented on crypto/ec/ec_elgamal_crypt.c:
```diff
@@ -0,0 +1,477 @@
+        if (len <= 0)
+            goto err;
+
+        if (!EC_POINT_point2oct(key->group, EC_GROUP_get0_generator(key->group),
```
buf需要改成依据len动态分配，否则可能导致溢出

wa5i replied:
> buf需要改成依据len动态分配，否则可能导致溢出

fixed

dongbeiouba commented on crypto/ec/ec_elgamal_crypt.c:
```diff
@@ -0,0 +1,477 @@
+        }
+
+        ERR_clear_error();
+    } while (1);
```
得看看是否可能导致死循环？

dongbeiouba commented on crypto/ec/ec_elgamal_crypt.c:
```diff
@@ -0,0 +1,477 @@
+    return NULL;
+}
+
+EC_ELGAMAL_CTX *EC_ELGAMAL_CTX_new_ex(EC_KEY *key, int32_t flag)
```
接口名字有歧义，因为绝大多数以_ex结尾的接口都是在原接口基础上增加参数，OSSL_LIB_CTX *libctx, const char *propq，是否考虑改一下接口名字和实现？例如改成EC_ELGAMAL_CTX_set_flags，（需要先调用EC_ELGAMAL_CTX_new()），

wa5i replied:
> 接口名字有歧义，因为绝大多数以_ex结尾的接口都是在原接口基础上增加参数，OSSL_LIB_CTX *libctx, const char *propq，是否考虑改一下接口名字和实现？例如改成EC_ELGAMAL_CTX_set_flags，（需要先调用EC_ELGAMAL_CTX_new()），

鉴于目前还没发布新版本，所以就直接在老接口上加了参数，防止接口过多不好维护，虽然在 master 分支上有一定的兼容性问题，但 ec-elgamal 目前用的不多，应该可控。

dongbeiouba commented on crypto/ec/ec_elgamal_encode.c:
```diff
@@ -0,0 +1,198 @@
+    if (size < len)
+        goto end;
+
+    if (ciphertext == NULL || ciphertext->C1 == NULL || ciphertext->C2 == NULL)
```
挪到函数入口处检查

wa5i replied:
> 挪到函数入口处检查

已修复

dongbeiouba commented on crypto/ec/ec_elgamal_encode.c:
```diff
@@ -0,0 +1,198 @@
+    unsigned char *p = in;
+    BN_CTX *bn_ctx = NULL;
+
+    if (ctx == NULL || ctx->key == NULL || r == NULL || r->C1 == NULL ||
```
in也需要判空

wa5i replied:
> in也需要判空

已修复

dongbeiouba commented on crypto/ec/ec_elgamal_encode.c:
```diff
@@ -0,0 +1,198 @@
+    if (size == (point_len + 1)) {
+        if (in[0] == 0) {
+            c1_len = 1;
+        } else if (in[point_len] == 0) {
```
point_len对应的POINT_CONVERSION_COMPRESSED的EC_POINT，如果EC_POINT编码是POINT_CONVERSION_UNCOMPRESSED，这里这么判断就错了吧。需要处理不同编码方式的情况。

wa5i replied:
> point_len对应的POINT_CONVERSION_COMPRESSED的EC_POINT，如果EC_POINT编码是POINT_CONVERSION_UNCOMPRESSED，这里这么判断就错了吧。需要处理不同编码方式的情况。

已修复

dongbeiouba commented on include/openssl/ec.h:
```diff
@@ -1948,6 +1965,17 @@ size_t EC_ELGAMAL_CIPHERTEXT_encode(EC_ELGAMAL_CTX *ctx, unsigned char *out,
+ *  \param  len     length of the string
+ *  \return 1 on success and 0 if an error occurred
+ */
+int EC_POINT_from_string(const EC_GROUP *group, EC_POINT *r,
```
EC_POINT_from_string()有测试用例吗？

wa5i replied:
> EC_POINT_from_string()有测试用例吗？

已加

dongbeiouba commented on crypto/ec/ec_elgamal_encode.c:
```diff
@@ -0,0 +1,178 @@
+
+    p += point_len;
+
+    plen = EC_POINT_point2oct(ctx->key->group, ciphertext->C2, form, p,
```
EC_POINT_point2oct()失败时返回0，建议改成plen == 0

wa5i replied:
> EC_POINT_point2oct()失败时返回0，建议改成plen == 0

fixed

dongbeiouba commented on crypto/ec/ec_elgamal_encode.c:
```diff
@@ -0,0 +1,178 @@
+
+    plen = EC_POINT_point2oct(ctx->key->group, ciphertext->C1, form, p,
+                              point_len, bn_ctx);
+    if (plen != 1 && plen != point_len)
```
EC_POINT_point2oct()失败时返回0，建议改成plen == 0

wa5i replied:
> EC_POINT_point2oct()失败时返回0，建议改成plen == 0

fixed

dongbeiouba commented on crypto/ec/ec_elgamal_encode.c:
```diff
@@ -0,0 +1,178 @@
+
+    plen = EC_POINT_point2oct(ctx->key->group, ciphertext->C1, form, p,
+                              point_len, bn_ctx);
+    if (plen != 1 && plen != point_len)
```
EC_POINT_point2oct()失败时返回0，建议改成plen == 0

wa5i replied:
> EC_POINT_point2oct()失败时返回0，建议改成plen == 0

fixed

dongbeiouba commented on crypto/ec/ec_elgamal_dlog.c:
```diff
@@ -0,0 +1,487 @@
+            break;
+    }
+
+    if (i >= max)
```
建议增加ERR_raise()

wa5i replied:
> 建议增加ERR_raise()

fixed

wa5i committed:
Add EC_POINT_from_string test case

wa5i committed:
put_str: Use memcpy instead of strncpy

This fixes a warning from latest gcc.

There is no point in using strncpy here as we
intentionally copy only the string contents without
the terminating NUL. The len is set from strlen().

dongbeiouba commented:
CI有失败 @wa5i

wa5i committed:
ec-elgamal test case bugfix

InfoHunter commented:
另外Twisted ElGamal在yuque上没有API文档，需要按照API文档的模板填写

wa5i commented:
> 另外Twisted ElGamal在yuque上没有API文档，需要按照API文档的模板填写

Twisted ElGamal 跟 EC-ElGamal 用法一样，只是参数不一样，多传了个 flag， 只新增了个API，回头我添加一下文档。

InfoHunter commented:
我看 EC ElGamal也没有：https://www.yuque.com/tsdoc/api/pk4xwr 可以一并处理了
